snowplow-tracker/src/main/java/com/snowplowanalytics/core/tracker/Tracker.kt

```diff
@@ -0,0 +1,828 @@
+            val eventId = event.eventId.toString()
+            val eventTimestamp = event.timestamp
+            val sessionManager = session
+            if (sessionManager == null) {
+                track(
+                    TAG,
+                    "Session not ready or method getHasLoadedFromFile returned false with eventId: %s",
+                    eventId
+                )
+                return
+            }
+            val sessionContextJson =
+                sessionManager.getSessionContext(eventId, eventTimestamp, userAnonymisation)
+            sessionContextJson?.let { event.contexts.add(it) }
+        }
+    }
+
+    private fun addBasicContextsToContexts(
+        contexts: MutableList<SelfDescribingJson>,
+        event: TrackerEvent
+    ) {
+        if (applicationContext) {
+            getApplicationContext(context)?.let { contexts.add(it) }
+        }
+        if (platformContextEnabled) {
+            platformContextManager.getMobileContext(userAnonymisation)?.let { contexts.add(it) }
+        }
+        if (event.isService) {
+            return
+        }
+        if (geoLocationContext) {
+            getGeoLocationContext(context)?.let { contexts.add(it) }
+        }
+        gdprContext?.let { contexts.add(it.context) }
+    }
+
+    private fun addGlobalContextsToContexts(
+        contexts: MutableList<SelfDescribingJson>,
+        event: InspectableEvent
+    ) {
+        synchronized(globalContextGenerators) {
+            for (generator in globalContextGenerators.values) {
+                contexts.addAll(generator.generateContexts(event))
+            }
+        }
+    }
+
+    private fun addStateMachineEntitiesToContexts(
+        contexts: MutableList<SelfDescribingJson>,
+        event: InspectableEvent
+    ) {
+        val stateManagerEntities = stateManager.entitiesForProcessedEvent(event)
+        contexts.addAll(stateManagerEntities)
+    }
+
+    private fun wrapContextsToPayload(payload: Payload, contexts: List<SelfDescribingJson>) {
+        if (contexts.isEmpty()) {
+            return
+        }
+        
+        val data: MutableList<Map<String, Any?>> = LinkedList()
+        for (context in contexts) {
+            data.add(context.map)
+        }
+        val finalContext = SelfDescribingJson(TrackerConstants.SCHEMA_CONTEXTS, data)
+        payload.addMap(
+            finalContext.map,
+            base64Encoded,
+            Parameters.CONTEXT_ENCODED,
+            Parameters.CONTEXT
+        )
+    }
+    
+    // --- Helpers
+    
+    /**
+     * Builds and adds a finalized payload of a service event
+     * by adding in extra information to the payload:
+     * - The event contexts (limited to identify the device and app)
+     * - The Tracker Subject
+     * - The Tracker parameters
+     *
+     * @param payload Payload the raw event payload to be
+     * decorated.
+     * @param contexts The raw context list
+     */
+    private fun addServiceEventPayload(
```

Comment:
I think we can remove this function, it is not used from anywhere and it is a private member so pretty safe to drop it. Probably just some relic of the past.